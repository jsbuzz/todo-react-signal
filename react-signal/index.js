import React from 'react';

import Control from './event-hive/control';
import StateConnector from './connect/StateConnector';

export const NamespaceCtx = React.createContext();

const connectComponent = ComponentClass => class extends ComponentClass {
  on = ns => Control.withActor(this, ns)

  namespace = () => Control.withActor(this, this.props.namespace)

  componentDidMount(...stuff) {
    super.componentDidMount && super.componentDidMount(...stuff);

    this.listen && this.listen();
    this.displayName = ComponentClass.name;
  }

  componentWillUnmount(...stuff) {
    super.componentWillUnmount && super.componentWillUnmount(...stuff);
    Control.cleanup(this);
  }
};

const connectFunction = (fn, namespace, events) => (props) => fn(
  props, (ns) => {
    ns || (ns = props.namespace);
    return Control.withActor(fn, ns)
  }
);

const Connect = (component, selector = null, namespace = null, events = null) => {
  let ConnectedComponent = Enable(component);

  if (selector) {
    ConnectedComponent = StateConnector(namespace, selector, events, ConnectedComponent)
  }

  const HiveComponent = (props) => (
    <NamespaceCtx.Consumer>
      { ctx => <ConnectedComponent {...props} namespace={ctx} /> }
    </NamespaceCtx.Consumer>
  );
  HiveComponent.displayName = `~${ConnectedComponent.displayName}`;

  return HiveComponent;
};

export const Enable = (component) => {
  let ConnectedComponent = component.prototype.render
    ? connectComponent(component)
    : connectFunction(component)
    ;

  ConnectedComponent.displayName = component.name;

  return ConnectedComponent;
};


export default Connect;
