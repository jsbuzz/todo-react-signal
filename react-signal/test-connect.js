import { EventGateway } from './event-hive/event-gateway';
import Control from './event-hive/control';

export class TestNameSpace extends EventGateway {
  events = [];
  lastEvent = null;

  namespaces = [];
  lastNamespace = [];

  trigger(event) {
    super.trigger(event);
    this.events.push(event);
    this.lastEvent = event;
  }

  withNamespace(NS) {
    this.namespaces.push(NS);
    this.lastNamespace = NS;
    return this;
  }

  reset() {
    this.events = [];
    this.lastEvent = null;
    this.namespaces = [];
    this.lastNamespace = null;
  }

  hasListenerFor(Event) {
    return this.eventPool.events.has(Event.EventName);
  }
};

// export const TestConnect = (fn, NS = new TestNameSpace()) => (props) => fn(
//   props, (ns) => (ns || NS)
// );

export const TestConnect = (ComponentClass, NS = new TestNameSpace()) => class extends ComponentClass {
  on = (ns) => NS.withNamespace(ns);
  namespace = () => NS.withNamespace(NS);

  // componentDidMount(...stuff) {
  //   super.componentDidMount && super.componentDidMount(...stuff);
  //
  //   this.listen && this.listen();
  //   this.displayName = ComponentClass.name;
  // }
};

export const StartService = (Service, NS) => {
  const service = new Service(NS);
  service.on = (ns) => NS.withNamespace(ns);
  service.namespace = () => NS.withNamespace(NS);
  Control.actor = service;
  service.listen();

  return service;
};
