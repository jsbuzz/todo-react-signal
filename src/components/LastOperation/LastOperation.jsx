import React from 'react';
import {
  AddTodo,
  RemoveTodo,
  UpdateTodo,
  RestoreTodos
} from '../../signal/events';

export const LastOperation = ({ operation, target }) => (
  <span>
    Last operation on {target}: <strong>{operation}</strong>
  </span>
);

export const LastOperationListeners = (get, set) => {
  const isTarget = ns => ns.name.includes(get('target'));
  return [
    AddTodo,
    ({ title, _origin }) =>
      isTarget(_origin) && set({ operation: `Added: "${title}"` }),

    RemoveTodo,
    ({ id, _origin }) =>
      isTarget(_origin) && set({ operation: `Removed: #${id}` }),

    UpdateTodo,
    ({ todo: { id, done }, _origin }) =>
      isTarget(_origin) &&
      set({
        operation: `Updated: #${id}: ${done ? 'done' : 'active'}`
      }),

    RestoreTodos,
    ({ savedTodos, _origin }) =>
      isTarget(_origin) &&
      set({
        operation: `Restored ${savedTodos.length} items from cache`
      })
  ];
};
