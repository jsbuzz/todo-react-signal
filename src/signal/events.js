import { Event } from '../../react-signal/event-hive/event';

export const AddTodoEvent = class extends Event {
  constructor(title) {
    super();
    this.title = title;
  }
};
export const IdEvent = class extends Event {
  constructor(id) {
    super();
    this.id = id;
  }
};

export const ErrorEvent = class extends Event {
  constructor(error) {
    super();
    this.error = error;
  }
};

export const AddTodo = AddTodoEvent.withAlias('Todo:add');
export const RemoveTodo = IdEvent.withAlias('Todo:remove');
export const ToggleTodo = IdEvent.withAlias('Todo:toggle');
