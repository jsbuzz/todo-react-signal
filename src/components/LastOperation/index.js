import { Signal } from '../../../react-signal';
import { LastOperationRenderer, LastOperationListeners } from './LastOperation';

export default Signal(LastOperationListeners)(LastOperationRenderer);
