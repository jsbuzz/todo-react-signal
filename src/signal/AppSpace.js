import {
  NameSpace,
  InitState,
  set,
} from '../../react-signal/event-hive/namespace';
import { UpdateActive } from './events';

const AppSpace = NameSpace.schema('Ns.App', () => ({
  activeTodos: [
    InitState, set(0),
    
    // update number of active todos
    UpdateActive, activeTodos => ({ active }) => console.log(active) || active
  ]
}));

window.AppSpace = AppSpace;

export default AppSpace;
