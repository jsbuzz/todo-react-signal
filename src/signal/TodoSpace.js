import {
  NameSpace,
  InitState,
  set,
  modify
} from '../../react-signal/event-hive/namespace';
import { AddTodo, RemoveTodo, UpdateTodo, RestoreTodos } from './events';

const TodoSpace = NameSpace.schema((todoId = 0) => ({
  todos: [
    InitState, set(new Map()),
    
    // Add new todo
    AddTodo, modify(todos => ({ title }) => {
      const id = ++todoId;
      const todo = { id, title, done: false };

      todos.set(id, todo);
    }),
    
    // restore todos from cache
    RestoreTodos, modify(todos => ({ savedTodos }) => {
      savedTodos.forEach(todo => {
        todoId = todo.id;
        todos.set(todo.id, todo);
      });
    }),
    
    // remove a todo by id
    RemoveTodo, modify(todos => ({ id }) => todos.delete(id)),
    
    // update a todo by id
    UpdateTodo, modify(todos => ({ todo }) => todos.set(todo.id, todo))
  ]
}));

window.TodoSpace = TodoSpace;

export default TodoSpace;
