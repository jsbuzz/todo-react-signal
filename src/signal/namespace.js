import { NameSpace, InitState } from '../../react-signal/event-hive/namespace';
import { AddTodo, RemoveTodo, ToggleTodo } from './events';

const TodoSpace = NameSpace.get('Ns.Todos');

const findTodo = (todos, todoId) => {
  let i = 0;
  for (const todo of todos) {
    if (todo.id === todoId) {
      return i;
    }
    i += 1;
  }
  return -1;
};

const toggleTodo = (todo) => {
  todo.done = !todo.done;
}

let todoId = 0;
TodoSpace.defineState({
  todos: state => [
    InitState, () => (state.todos = []),
    AddTodo, ({ title }) => (state.todos.push({
      id: ++todoId,
      title,
      done: false,
    })),
    RemoveTodo, ({ id }) => (state.todos.splice(findTodo(state.todos, id), 1)),
    ToggleTodo, ({ id }) => (toggleTodo(state.todos[findTodo(state.todos, id)])),
  ],
});

export default TodoSpace;
