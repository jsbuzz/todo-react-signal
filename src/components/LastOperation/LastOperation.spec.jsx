import React from 'react';
import { shallow } from 'enzyme';
import { TestNameSpace } from '../../../react-signal/test-connect';

import { AddTodo } from '../../signal/events';
import { LastOperationListeners, LastOperation } from './LastOperation';

describe('components/LastOperation', () => {
  const target = 'Target';
  let NS;
  let component;

  beforeEach(() => {
    NS = new TestNameSpace('TargetSpace');
    const SignalComponent = NS.addSignal(LastOperationListeners)(LastOperation);
    component = shallow(<SignalComponent target={target} />);
  });

  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  describe('WHEN AddTodo is triggered', () => {

  });
});