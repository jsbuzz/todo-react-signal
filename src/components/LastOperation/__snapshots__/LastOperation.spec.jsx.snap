// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/LastOperation should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <~LastOperation
    target="Target"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "namespace": TestNameSpace {
        "_parent": undefined,
        "_sendStateUpdates": true,
        "eventPool": EventPool {
          "allEvents": null,
          "events": Map {},
        },
        "events": Array [],
        "id": 2,
        "lastEvent": null,
        "logging": undefined,
        "name": "TargetSpace",
      },
      "target": "Target",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "namespace": TestNameSpace {
          "_parent": undefined,
          "_sendStateUpdates": true,
          "eventPool": EventPool {
            "allEvents": null,
            "events": Map {},
          },
          "events": Array [],
          "id": 2,
          "lastEvent": null,
          "logging": undefined,
          "name": "TargetSpace",
        },
        "target": "Target",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
